pkg/cmd/repo/autolink/list/list_test.go

```diff
@@ -0,0 +1,247 @@
+package list
+
+import (
+	"bytes"
+	"fmt"
+	"net/http"
+	"testing"
+
+	"github.com/MakeNowJust/heredoc"
+	"github.com/cli/cli/v2/internal/browser"
+	"github.com/cli/cli/v2/internal/ghrepo"
+	"github.com/cli/cli/v2/pkg/cmdutil"
+	"github.com/cli/cli/v2/pkg/iostreams"
+	"github.com/google/shlex"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestNewCmdList(t *testing.T) {
+	tests := []struct {
+		name         string
+		input        string
+		output       listOptions
+		wantErr      bool
+		wantExporter bool
+		errMsg       string
+	}{
+		{
+			name:   "no argument",
+			input:  "",
+			output: listOptions{},
+		},
+		{
+			name:   "web flag",
+			input:  "--web",
+			output: listOptions{WebMode: true},
+		},
+		{
+			name:         "json flag",
+			input:        "--json id",
+			output:       listOptions{},
+			wantExporter: true,
+		},
+		{
+			name:    "invalid json flag",
+			input:   "--json invalid",
+			wantErr: true,
+			errMsg: heredoc.Doc(`
+				Unknown JSON field: "invalid"
+				Available fields:
+				  id
+				  isAlphanumeric
+				  keyPrefix
+				  urlTemplate`),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ios, _, _, _ := iostreams.Test()
+			f := &cmdutil.Factory{
+				IOStreams: ios,
+			}
+			f.HttpClient = func() (*http.Client, error) {
+				return &http.Client{}, nil
+			}
+
+			argv, err := shlex.Split(tt.input)
+			require.NoError(t, err)
+
+			var gotOpts *listOptions
+			cmd := NewCmdList(f, func(opts *listOptions) error {
+				gotOpts = opts
+				return nil
+			})
+
+			cmd.SetArgs(argv)
+			cmd.SetIn(&bytes.Buffer{})
+			cmd.SetOut(&bytes.Buffer{})
+			cmd.SetErr(&bytes.Buffer{})
+
+			_, err = cmd.ExecuteC()
+			if tt.wantErr {
+				require.EqualError(t, err, tt.errMsg)
+			} else {
+				require.NoError(t, err)
+				assert.Equal(t, tt.output.WebMode, gotOpts.WebMode)
+				assert.Equal(t, tt.wantExporter, gotOpts.Exporter != nil)
+			}
+		})
+	}
+}
+
+type mockAutoLinkGetter struct {
+	Response []autolink
+}
+
+func (m *mockAutoLinkGetter) Get(repo ghrepo.Interface) ([]autolink, error) {
+	return m.Response, nil
+}
+
+type mockAutoLinkGetterError struct {
+	err error
+}
+
+func (me *mockAutoLinkGetterError) Get(repo ghrepo.Interface) ([]autolink, error) {
+	return nil, me.err
+}
+
+func TestListRun(t *testing.T) {
+	tests := []struct {
+		name       string
+		opts       *listOptions
+		isTTY      bool
+		response   []autolink
```

Comment:
**Suggestion**

In order to fix https://github.com/cli/cli/pull/10124/files#r1896710006 I'd encourage removing `response`, and the two "mocks" above and instead doing the following:

```go
type stubAutoLinkGetter struct {
	autolinks []autolink
        err error
}

func (g stubAutoLinkGetter) Get(repo ghrepo.Interface) ([]autolink, error) {
	return m.autolinks, m.err
}
```

And in the tests instantiate an instance:

```go
stubGetter: stubAutoLinkGetter{
   ...
}
```

Reply:
Agreed, that's much better 👍 Not sure why I didn't do that on my first pass 😅

Reply:
Implemented in da826db34227bb8aeb8d7e273a9abafcb6e7e748. Have a look!